frontend/src/plugins/impl/FileBrowserPlugin.tsx

```diff
@@ -0,0 +1,279 @@
+  const [path, setPath] = useState(initialPath);
+
+  const { data, loading, error } = useAsyncData(
+    () =>
+      list_directory({
+        path: path,
+        filetypes: filetypes,
+      }),
+    [path],
+  );
+
+  if (loading && !data) {
+    return null;
+  }
+
+  if (error) {
+    console.error(error);
+    toast({
+      title: `Could not load files in directory ${path}`,
+      description: error.message,
+      variant: "danger",
+    });
+  }
+
+  let { files } = data || {};
+  if (files === undefined) {
+    files = [];
+  }
+
+  // For checkbox logic
+  const selectedPaths = new Set(value.map((x) => x.path));
+
+  // For displaying selected files
+  const selectedFiles = value.map((x) => <li key={x.id}>{x.path}</li>);
+
+  /**
+   * Set new path from user input or selection.
+   * @param {string} newPath - New path
+   */
+  function setNewPath(newPath: string) {
+    // Navigate to parent directory
+    if (newPath === "..") {
+      newPath = path.endsWith("/") ? path.slice(0, -1) : path;
+      newPath = newPath.slice(0, Math.max(0, newPath.lastIndexOf("/")));
+    }
+
+    // If restricting navigation, check if path is outside bounds
+    const outsideInitialPath = newPath.length < initialPath.length;
+    if (restrictNavigation && outsideInitialPath) {
+      toast({
+        title: "Access denied",
+        description:
+          "Access to directories outside initial path is restricted.",
+        variant: "danger",
+      });
+      return;
+    }
+
+    setPath(newPath);
+  }
+
+  /**
+   * Handles file selection.
+   * @param {string} path - Path of selected file
+   * @param {string} name - Name of selected file
+   */
+  const selectFile = (path: string, name: string) => {
+    const fileInfo: FileInfo = {
+      id: path,
+      name: name,
+      path: path,
+      is_directory: false,
+      is_marimo_file: false,
+    };
+
+    if (multiple) {
+      if (selectedPaths.has(path)) {
+        setValue(value.filter((x) => x.path !== path));
+      } else {
+        setValue([...value, fileInfo]);
+      }
+    } else {
+      setValue([fileInfo]);
+    }
+  };
+
+  const fileRows = [];
+
+  // Parent directory ".." row button
+  fileRows.push(
+    <TableRow key={"Parent directory"} onClick={() => setNewPath("..")}>
+      <TableCell className="w-1/12" />
+      <TableCell className="w-11/12">..</TableCell>
+    </TableRow>,
+  );
+
+  // Create rows for directories and files
+  for (const file of files) {
+    let filePath = path;
+    if (!path.endsWith("/")) {
+      filePath += "/";
+    }
+    filePath += file.name;
+
+    const handleClick = file.is_directory ? setNewPath : selectFile;
+
+    fileRows.push(
+      <TableRow key={file.id} onClick={() => handleClick(filePath, file.name)}>
+        <TableCell className="w-1/12">
+          {file.is_directory ? (
+            <Folder size={16} className="ml-2" />
+          ) : (
+            <Checkbox className="ml-2" checked={selectedPaths.has(filePath)} />
```

Comment:
if you wanted to show icons for different file types (not necessary, but if you want to design that), we have utils `guessFileType` `FILE_TYPE_ICONS` to get an icon for a given extension

Reply:
Oh neat, let me look into that